pkg/component/component.go

```diff
@@ -120,6 +120,164 @@ func ApplyConfig(client kclient.ClientInterface, envSpecificInfo envinfo.EnvSpec
+				// If it's blank, and there is an instance of the same name with a managed-by label, we propagate that to the output
+				// components
+				if componentLabel[componentlabels.KubernetesManagedByLabel] == "" && checkLabel[componentlabels.KubernetesManagedByLabel] != "" {
+					component.ObjectMeta.Labels[componentlabels.KubernetesManagedByLabel] = checkLabel[componentlabels.KubernetesManagedByLabel]
+				}
+
+				// Check if there are TWO components, one in dev, one in deploy. If so, we will keep the *first* component, and
+				// update the mode label to indicate it's running in both dev and deploy.
+				// We check this *explicitely* in order to keep the correct order when displaying (ex. Dev, Deploy)
+				// instead of some components showing Deploy, Dev depending on the order that the components are in the array.
+				// We also use this long if statement in case there is something odd that went on (user modified the label outside of odo)
+				componentMode := componentLabel[componentlabels.OdoModeLabel]
+				checkMode := checkLabel[componentlabels.OdoModeLabel]
+				if (componentMode != checkMode) &&
+					(componentMode == componentlabels.ComponentDevName || componentMode == componentlabels.ComponentDeployName) &&
+					(checkMode == componentlabels.ComponentDevName || checkMode == componentlabels.ComponentDeployName) {
+					component.ObjectMeta.Labels[componentlabels.OdoModeLabel] = fmt.Sprintf("%s, %s", componentlabels.ComponentDevName, componentlabels.ComponentDeployName)
+				}
+
+				// Some components have no annotations (ex. knative), so must propagage the ObjectMeta with initial annotations.
+				if component.ObjectMeta.Annotations == nil {
+					component.ObjectMeta.Annotations = make(map[string]string)
+				}
+
+				// Check to see if the next component has a project type annotation, and update the component accordingly if it does.
+				if componentAnnotation[componentlabels.OdoProjectTypeAnnotation] == "" && checkAnnotation[componentlabels.OdoProjectTypeAnnotation] != "" {
+					component.ObjectMeta.Annotations[componentlabels.OdoProjectTypeAnnotation] = checkAnnotation[componentlabels.OdoProjectTypeAnnotation]
+				}
+
+			}
+		}
+
+		// Go through the output list, and add the component if it hasn't already been added.
+		found := false
+		for _, outputComponent := range outputComponents {
+			// Go by component name instead of instance-of
+			if outputComponent.Name == component.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			outputComponents = append(outputComponents, component)
+		}
+
+	}
+	return outputComponents
+}
+
+// ListAllClusterComponents returns a list of all "components" on a cluster
+// that are both odo and non-odo components.
+//
+// We then return a list of "components" intended for listing / output purposes specifically for commands such as:
+// `odo list`
+func ListAllClusterComponents(client kclient.ClientInterface, namespace string) ([]Component, error) {
```

Comment:
This function returns a list of Kubernetes resources, which is not what is intended IMO. 

It should instead return a slice of struct similar to 

```
OdoComponent {
- Name
- Managed By
- Mode
- Type
}
```

which represent an odo component, which can be implemented by several Kubernetes resources. IMO we need to get rid of the details (Kubernetes resources) and only return the high level object (the Odo component).

For this, you can start with an empty slice of OdoComponent, and you go through the Kubernetes resources, from which you extract the Name / Managed-by / Mode/type from Labels/annotations, and if such an OdoComponent doesn't exist on the list of OdoComponent, you add it to the list

Reply:
Yeah, you are right. I'll refactor it so it does most of the logic here, instead of spitting out *every* Kubernetes resource.